File: lua/all-your-base/all-your-base.lua
--- A module for converting numbers between different bases.
local all_your_base <const> = {}

--- Converts a list of digits from a given base to a single decimal value.
--- This function validates the input base and all digits. It will throw an error if:
--- - The `base` is less than 2.
--- - A `digit` is negative.
--- - A `digit` is greater than or equal to the `base`.
--- @param digits integer[] A table of digits to be converted.
--- @param base integer The base of the input digits.
--- @return integer -- The decimal value of the number.
local function to_value(digits, base)
    assert(base >= 2, "invalid input base")

    local digits_count <const> = #digits
    local value = 0

    for i = digits_count, 1, -1 do
        local digit <const> = digits[i]

        assert(digit >= 0, "negative digits are not allowed")
        assert(digit < base, "digit out of range")

        value = value + digit * base ^ (digits_count - i)
    end

    return value
end

--- Converts a decimal value to a list of digits in a new base.
--- This function validates the output base. It will throw an error if the `base` is less than 2.
--- @param value integer The decimal value to convert.
--- @param base integer The base to convert to.
--- @return integer[] -- A table of digits representing the number in the new base.
local function from_value(value, base)
    assert(base >= 2, "invalid output base")

    if value == 0 then
        return { 0 }
    end

    local digits <const> = {}
    local current = value

    while current > 0 do
        table.insert(digits, 1, current % base)
        current = math.floor(current / base)
    end

    return digits
end

--- Initializes the base conversion process.
--- This function takes the number to be converted in its initial base and returns a table with a `to` method to complete the conversion.
--- @param input_digits integer[] A table of digits to be converted.
--- @param input_base integer The base of the input digits.
--- @return {to: fun(base: integer): integer} -- A table containing the final conversion method.
function all_your_base.convert(input_digits, input_base)
    local value = to_value(input_digits, input_base)

    return {
        --- The final conversion method.
        --- This function converts the number to the final desired base.
        --- It returns a copy of the input digits if the bases are the same.
        --- @param base integer The base to convert to.
        --- @return integer[] -- A table of digits representing the number in the new base.
        to = function(base)
            if input_base == base then
                return { table.unpack(input_digits) }
            end
            return from_value(value, base)
        end
    }
end

return all_your_base

File: go/all-your-base/all_your_base.go
// Package allyourbase provides functions for converting numbers between arbitrary bases.
package allyourbase

import (
	"errors"
	"math"
)

// ConvertToBase converts a number represented by `inputDigits` in the `inputBase`
// to a new representation in the `outputBase`.
//
// It returns a slice of integers with the digits in the new base or an error if
// the bases are invalid or if input digits are outside the valid range.
func ConvertToBase(inputBase int, inputDigits []int, outputBase int) ([]int, error) {
	if inputBase < 2 {
		return conversionFail("input base must be >= 2")
	}
	if outputBase < 2 {
		return conversionFail("output base must be >= 2")
	}
	if !validateNumber(inputDigits, inputBase) {
		return conversionFail("all digits must satisfy 0 <= d < input base")
	}
	if len(inputDigits) == 0 {
		return conversionSuccess([]int{0})
	}

	numberValue := desconvertFromBase(inputDigits, inputBase)
	resultValue := convertToBase(numberValue, outputBase)

	return conversionSuccess(resultValue)
}

// validateNumber checks if all digits in the slice are within the valid range
// for the given base (0 <= d < base).
func validateNumber(digits []int, base int) bool {
	for _, digit := range digits {
		if digit < 0 || digit >= base {
			return false
		}
	}

	return true
}

// desconvertFromBase calculates the numerical magnitude of the number represented
// by the input digits in an arbitrary base.
//
// It returns this magnitude as a standard Go integer value.
func desconvertFromBase(digits []int, base int) int {
	totalDigits := len(digits)
	value := 0

	for index, digit := range digits {
		position := totalDigits - index - 1
		value += valueOfDigit(digit, position, base)
	}

	return value
}

// convertToBase converts an integer value into a slice of digits
// representing that number in the target arbitrary base.
//
// It uses the method of successive division.
func convertToBase(value, base int) []int {
	if value == 0 {
		return []int{0}
	}

	currentDigits := []int{}

	for currentValue := value; currentValue > 0; currentValue /= base {
		digit := currentValue % base
		currentDigits = append(currentDigits, digit)
	}

	return reverseSlice(currentDigits)
}

// valueOfDigit calculates the weighted value of a single digit based on its position and base.
func valueOfDigit(digit, position, base int) int {
	return digit * intPow(base, position)
}

// conversionFail is a utility function to return a failure result with an empty
// slice and a formatted error.
func conversionFail(message string) ([]int, error) {
	return []int{}, errors.New(message)
}

// conversionSuccess is a utility function to return a successful result with the
// converted value and a nil error.
func conversionSuccess(value []int) ([]int, error) {
	return value, nil
}

// intPow calculates the integer power x^y.
func intPow(x, y int) int {
	fX := float64(x)
	fY := float64(y)

	return int(math.Pow(fX, fY))
}

// reverseSlice reverses the order of elements in an integer slice.
func reverseSlice(source []int) []int {
	size := len(source)
	reversed := make([]int, size)

	for index, value := range source {
		revIndex := size - index - 1
		reversed[revIndex] = value
	}

	return reversed
}
